driver/generic/st77xx.py

```diff
@@ -0,0 +1,415 @@
+ST77XX_MADCTL_ML = const(0x10)  # line address order (0: refresh to to bottom; 1: refresh bottom to top)
+ST77XX_MADCTL_BGR = const(0x08) # colors are BGR (not RGB)
+ST77XX_MADCTL_RTL = const(0x04) # refresh right to left
+ST77XX_MADCTL_ROTS=(
+    const(0x00),                                # 0 = portrait
+    const(ST77XX_MADCTL_MX | ST77XX_MADCTL_MV), # 1 = landscape
+    const(ST77XX_MADCTL_MY | ST77XX_MADCTL_MX), # 2 = inverted portrait
+    const(ST77XX_MADCTL_MY | ST77XX_MADCTL_MV), # 3 = inverted landscape
+)
+
+ST77XX_COLOR_MODE_65K = const(0x50)
+ST77XX_COLOR_MODE_262K = const(0x60)
+ST77XX_COLOR_MODE_12BIT = const(0x03)
+ST77XX_COLOR_MODE_16BIT = const(0x05)
+ST77XX_COLOR_MODE_18BIT = const(0x06)
+ST77XX_COLOR_MODE_16M = const(0x07)
+
+
+ST77XX_COL_ROW_MODEL_START_ROTMAP={
+    # ST7789
+    (240,320,None):[(0,0),(0,0),(0,0),(0,0)],
+    (240,240,None):[(0,0),(0,0),(0,80),(80,0)],
+    (135,240,None):[(52,40),(40,53),(53,40),(40,52)],
+    # ST7735
+    (128,160,'blacktab'):[(0,0),(0,0),(0,0),(0,0)],
+    (128,160,'redtab'):[(2,1),(1,2),(2,1),(1,2)],
+}
+
+ST77XX_PORTRAIT = const(0)
+ST77XX_LANDSCAPE = const(1)
+ST77XX_INV_PORTRAIT = const(2)
+ST77XX_INV_LANDSCAPE = const(3)
+
+class St77xx_hw(object):
+    def __init__(self, *, cs, dc, spi, res, suppRes, bl=None, model=None, suppModel=[], rst=None, rot=ST77XX_LANDSCAPE, bgr=False, rp2_dma=None):
+        '''
+        This is an abstract low-level driver the ST77xx controllers, not to be instantiated directly.
+        Derived classes implement chip-specific bits. THe following parameters are recognized:
+
+        * *cs*: chip select pin (= slave select, SS)
+        * *dc*: data/command pin
+        * *bl*: backlight PWM pin (optional)
+        * *model*: display model, to account for variations in products
+        * *rst*: optional reset pin
+        * *res*: resolution tuple; (width,height) with zero rotation
+        * *rot*: display orientation (0: portrait, 1: landscape, 2: inverted protrait, 3: inverted landscape); the constants ST77XX_PORTRAIT, ST77XX_LANDSCAPE, ST77XX_INV_POTRAIT, ST77XX_INV_LANDSCAPE may be used.
+        * *bgr*: color order if BGR (not RGB)
+        * *rp2_dma*: optional DMA object for the rp2 port
+
+
+        Subclass constructors (implementing concrete chip) set in addition the following, not to be used directly:
+
+        * *suppModel*: models supported by the hardware driver
+        * *suppRes*: resolutions supported by the hardware driver, as list of (width,height) tuples
+        '''
+        self.buf1 = bytearray(1)
+        self.buf2 = bytearray(2)
+        self.buf4 = bytearray(4)
+
+        self.cs,self.dc,self.bl,self.rst=[(machine.Pin(p,machine.Pin.OUT) if isinstance(p,int) else p) for p in (cs,dc,bl,rst)]
+        self.bl.value(1)
+        self.bl=machine.PWM(self.bl)
+        self.rot=rot
+        self.bgr=bgr
+        self.width,self.height=(0,0) # set in apply_rotation
```

Comment:
What is the motivation in calling `apply_rotation` on each subclass instead of once on the base class? (probably right after calling `config`)

Reply:
The motivation was that apply_rotation manipulates the registers, thus the logical place for it was at the end of each subclass's `config` which does the first part of the hardware setup. I could also create a method `St77xx_hw.config` calling the real init sequence as e.g. `St7735_hw.config_hw`  and then `St77xx.apply_rotation`.

Reply:
If you always set these registers the same way on all hw flavours, I don't see the benefit in duplicating code.  

In fact, you could simply pass the init array to the St77xx_hw constructor (in `super().__init__` ) and let St77xx_hw call both `_run_seq` and `apply_rotation`, moving all functionality to St77xx_hw and keeping only configuration data on the derived classes.

Reply:
I created `config_hw` in child classes based on my previous suggestion. Putting everything to St77xx_hw has its pros, but also cons: it will duplicate supported resolution / model handling in the child class ctors (perhaps others).